[chore] Upgrade NX from v17 to v22

ggdouglas opened the pull request:
## Summary
Upgrades NX from `17.2.8` to `22.3.1` through incremental major version migrations (17→18→19→20→21→22).

## Changes
- **nx.json**:
  - Removed deprecated `tasksRunnerOptions`
  - Moved `affected.defaultBase` to root-level `defaultBase`
  - Added `useInferencePlugins: false to maintain` explicit target configuration
  - Added `continuous: true` to `dev` target for NX 22 compatibility with watch tasks
- **package.json**: Excluded `demo-app` from default `dev` script to avoid port 9001 conflict with `docs-app`
- **.gitignore / .prettierignore**: Added NX workspace-data directory ignores

## Testing
- `pnpm run verify` passes (compile, dist, test, lint, format-check)
- `pnpm dev` runs successfully with continuous watch tasks

## Demo

https://github.com/user-attachments/assets/c4410dd0-fec5-41d2-8b25-5dd9cdf5e82a

ggdouglas committed:
chore: upgrade NX 17 to 18

- Updated nx from 17.2.8 to 18.3.5
- Added useInferencePlugins: false to disable Project Crystal
- Moved affected.defaultBase to root level defaultBase

ggdouglas committed:
chore: upgrade NX 19 to 20

- Updated nx from 19.8.9 to 20.8.3
- Removed deprecated tasksRunnerOptions
- Added useLegacyCache: true for migration

ggdouglas committed:
chore: upgrade NX 20 to 21

- Updated nx from 20.8.3 to 21.6.9
- Removed useLegacyCache (no longer needed)
- Updated .gitignore for nx rule files

ggdouglas committed:
chore: upgrade NX 21 to 22.3.1

- Updated nx from 21.6.9 to 22.3.1

svc-palantir-github commented:
<h3>chore: upgrade NX 21 to 22.3.1</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/a5593454-6879-452a-8b2a-af6aeb2bc800/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/a5593454-6879-452a-8b2a-af6aeb2bc800/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/a5593454-6879-452a-8b2a-af6aeb2bc800/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/a5593454-6879-452a-8b2a-af6aeb2bc800/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
fix: add continuous flag to dev target for NX 22 compatibility

NX 22 requires long-running watch tasks to be marked with continuous: true
so that NX knows not to wait for them to complete.

ggdouglas committed:
fix: exclude demo-app from default dev script to avoid port conflict

Both docs-app and demo-app use port 9001 by default. Running them
in parallel causes EADDRINUSE errors.

ggdouglas committed:
chore: upgrade NX 18 to 19

- Updated nx from 18.3.5 to 19.8.9
- Updated workspace data directory config

svc-palantir-github commented:
<h3>fix: exclude demo-app from default dev script to avoid port conflict</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/8011a1f4-be00-4b83-bfc0-5a55e26da7b2/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/8011a1f4-be00-4b83-bfc0-5a55e26da7b2/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/8011a1f4-be00-4b83-bfc0-5a55e26da7b2/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/8011a1f4-be00-4b83-bfc0-5a55e26da7b2/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

svc-palantir-github commented:
<h3>fix: exclude demo-app from default dev script to avoid port conflict</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/ed17d252-1e3f-4f55-bda9-7ed840aca987/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/ed17d252-1e3f-4f55-bda9-7ed840aca987/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/ed17d252-1e3f-4f55-bda9-7ed840aca987/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/ed17d252-1e3f-4f55-bda9-7ed840aca987/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas commented on nx.json:
```diff
@@ -1,12 +1,4 @@
 {
-    "affected": {
-        "defaultBase": "develop"
```
Removed `affected.defaultBase` → Moved to root `defaultBase`
```diff
-    "affected": {
-        "defaultBase": "develop"
-    },
+    "defaultBase": "develop"
```

- [`move-default-base-to-nx-json-root`](https://nx.dev/docs/reference/nx/migrations#move-default-base-to-nx-json-root) (Nx 18.1.0-beta.3)
- https://nx.dev/docs/reference/nx-json#default-base

ggdouglas commented on nx.json:
```diff
@@ -1,12 +1,4 @@
-    "affected": {
-        "defaultBase": "develop"
-    },
-    "tasksRunnerOptions": {
```
Removed `tasksRunnerOptions`
```diff
-    "tasksRunnerOptions": {
-        "default": {
-            "runner": "nx/tasks-runners/default"
-        }
-    },
```

- [`remove-custom-tasks-runner`](https://nx.dev/docs/reference/nx/migrations#remove-custom-tasks-runner) (Nx 21.0.0-beta.8)
- [Legacy Cache / Custom Task Runners Deprecation](https://nx.dev/deprecated/custom-tasks-runner)
- `tasksRunnerOptions` deprecated in Nx 20, removed in Nx 21

ggdouglas commented on nx.json:
```diff
@@ -24,7 +16,9 @@
-            "dependsOn": ["^compile"]
+            "dependsOn": ["^compile"],
+            "continuous": true,
+            "cache": false
```
Added `continuous: true` and `cache: false` to dev target:

Nx 21 introduced continuous tasks for long running processes like dev servers. Without `continuous: true`, tasks that depend on `dev` will wait indefinitely for it to complete. 21aa7d9d3a0d0a51b2d25a3702d9c0f99ebe6c7b

- [Nx 21 Release: Continuous Tasks](https://nx.dev/blog/nx-21-release)
- [Continuous Tasks Deep Dive](https://nx.dev/blog/nx-21-continuous-tasks)

ggdouglas commented on nx.json:
```diff
@@ -106,5 +100,7 @@
         "production": ["default"]
-    }
+    },
+    "useInferencePlugins": false,
```
Added `useInferencePlugins: false`

When upgrading existing workspaces, Nx sets this to false to maintain explicit target configuration rather than auto inferring targets from config files.

- [`18.0.0-disable-adding-plugins-for-existing-workspaces`](https://nx.dev/docs/reference/nx/migrations#1800-disable-adding-plugins-for-existing-workspaces) (Nx 18.0.0-beta.2)
- [Inferred Tasks (Project Crystal) > Existing Nx Workspaces](https://nx.dev/docs/concepts/inferred-tasks#existing-nx-workspaces)

svc-palantir-github commented:
<h3>fix: exclude demo-app from default dev script to avoid port conflict</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/f170b42e-e012-4d63-ad69-7311fb022ca4/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/f170b42e-e012-4d63-ad69-7311fb022ca4/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/f170b42e-e012-4d63-ad69-7311fb022ca4/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/f170b42e-e012-4d63-ad69-7311fb022ca4/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Include demo-app in dev command by running on separate port (9002)

svc-palantir-github commented:
<h3>Include demo-app in dev command by running on separate port (9002)</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/ae9cf951-682a-4fec-989e-60114c6a9e44/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/ae9cf951-682a-4fec-989e-60114c6a9e44/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/ae9cf951-682a-4fec-989e-60114c6a9e44/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/ae9cf951-682a-4fec-989e-60114c6a9e44/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

mm-wang reviewed (approved):
LGTM!
